### PR TITLE
Add Trigger and Subscription modal in topology

### PR DESCRIPTION
**Pull request by karthikjeeyar**
**Fixes:** https://issues.redhat.com/browse/ODC-4246 , https://issues.redhat.com/browse/ODC-4167

**Problem:**
Ability to add trigger/subscription from topology gestures and kebab actions

**Solution:**
Added Trigger modal and kebab actions for triggers.
Added Subscription modal and kebab actions for subscriptions
Added context menu on the resource backed edges

**Note: Sidebar enhancements will be added in a follow up PR**

**Screenshot:**
Trigger Modal:

image

Subscription Modal:

image

Empty state:

image

Error state:
image

Gif:
brokers-triggers-modal

cc: @rachael-phillips @beaumorley @openshift/team-devconsole-ux

### CODE REVIEW

**Comment by serenamarie125** [comment]
@rachael-phillips can you confirm?  I think that the label on the Add Trigger modal should be "Subscriber", not "Select Subscriber"

**Comment by serenamarie125** [comment]
Can we have the Attribute/Value <delete icon> be the same width as the other input widgets?  Right now there's extra space between the delete icon & the edge of the modal.

**Comment by karthikjeeyar** [comment]
> @rachael-phillips can you confirm? I think that the label on the Add Trigger modal should be "Subscriber", not "Select Subscriber"

@serenamarie125 @rachael-phillips  Like we have in Design doc, I have changed the title to `Subscriber` and updated the screenshots. The filter attributes section in the trigger modal uses the `nameValueEditor` field that we use every where in console has this type of spacing.
eg: ApiServersource, sinkbinding event source forms

image

**Comment by rachael-phillips** [comment]
Hey @karthikjeeyar @serenamarie125 I can confirm that the recommendation to have Subscriber rather than Select Subscriber is correct. I think we should eventually update the spacing of the Resource form fields across the flow, but agree with keeping it standard for now rather than updating based on what I've seen in other places. For example, this is another modal on the topology view, and for consistency I think we should go with the standard spacing. 
Screen Shot 2020-07-23 at 9 50 05 AM

@serenamarie125 I will add your recommendation to this Audit I am currently working on of our form fields so we can address this in the future: https://docs.google.com/document/d/10QMFBrCGmwbq92Vy8GSNteAX6EgslTcR4PnMoQwF_1w/edit?usp=sharing

File: frontend/packages/dev-console/src/components/topology/components/edges/BaseEdge.tsx [hunk 1]
```diff
@@ -17,7 +18,8 @@ type BaseEdgeProps = {
   element: Edge;
   dragging?: boolean;
   className?: string;
-} & WithRemoveConnectorProps;
+} & WithRemoveConnectorProps &
+  WithContextMenuProps;
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/components/edges/BaseEdge.tsx:22]: If you use `Partial` then the components that use the `BaseEdge` do not necessarily need to provide context menu props.
```suggestion
  Partial<WithContextMenuProps>;
```

File: frontend/packages/dev-console/src/components/topology/components/edges/TrafficConnector.tsx [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
 import './TrafficConnector.scss';
 
 type TrafficConnectorProps = {
   element: Edge;
-};
+} & WithContextMenuProps;
 
-const TrafficConnector: React.FC<TrafficConnectorProps> = ({ element }) => (
-  <BaseEdge element={element} className="odc-traffic-connector">
+const TrafficConnector: React.FC<TrafficConnectorProps> = ({ element, ...others }) => (
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/components/edges/TrafficConnector.tsx:10]: Shouldn't need to change this component if `BaseEdge` props are defined properly.

File: frontend/packages/knative-plugin/src/actions/add-subscription.tsx [hunk 3]
```diff
@@ -15,3 +21,9 @@ export const addSubscription = (model: K8sKind, source: K8sResourceKind): KebabO
     },
   };
 };
+
+export const addSubscriptionWithIcon = (source): KebabOption => {
+  const subscriptionMenu = addSubscription(EventingSubscriptionModel, source);
+  subscriptionMenu.icon = <StarIcon />;
```
Comment (christianvogt) [inline frontend/packages/knative-plugin/src/actions/add-subscription.tsx:27]: Remove the star icon. Fairly certain it was agreed now to have no icons.

cc @openshift/team-devconsole-ux

File: frontend/packages/knative-plugin/src/actions/add-trigger.tsx [hunk 4]
```diff
@@ -15,3 +21,9 @@ export const addTrigger = (model: K8sKind, source: K8sResourceKind): KebabOption
     },
   };
 };
+
+export const addTriggerWithIcon = (source): KebabOption => {
+  const triggerMenu = addTrigger(EventingTriggerModel, source);
+  triggerMenu.icon = <StarIcon />;
```
Comment (christianvogt) [inline frontend/packages/knative-plugin/src/actions/add-trigger.tsx:27]: Same comment.
No icon.

File: frontend/packages/knative-plugin/src/components/pub-sub/form-fields/PubSubSubscriber.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  const handleOnLoad = (resourceList: { [key: string]: string }) =>
+    setResourceAlert(_.isEmpty(resourceList));
+
+  const dropdownResources = knativeServingResourcesServices(values.metadata.namespace);
+  return (
+    <FormGroup fieldId={getFieldId('pubsub', 'subscriber')} label="Subscriber" isRequired>
+      {resourceAlert && (
+        <>
+          <Alert variant="default" title="No Subscriber available" isInline>
+            To create a Subscriber create a Knative Service from the Add page.
```
Comment (christianvogt) [inline frontend/packages/knative-plugin/src/components/pub-sub/form-fields/PubSubSubscriber.tsx:42]:
```suggestion
            To create a Subscriber create, a Knative Service from the Add page.
```

File: frontend/packages/knative-plugin/src/topology/components/edges/EventSourceLink.tsx [hunk 6]
```diff
@@ -16,7 +17,8 @@ type EventSourceLinkProps = {
   element: Edge;
   dragging: boolean;
 } & WithSourceDragProps &
-  WithTargetDragProps;
+  WithTargetDragProps &
+  WithContextMenuProps;
```
Comment (christianvogt) [inline frontend/packages/knative-plugin/src/topology/components/edges/EventSourceLink.tsx:21]: Changes here aren't necessary if you type `BaseEdge` correctly.

File: frontend/packages/knative-plugin/src/topology/components/edges/TrafficLink.tsx [hunk 7]
```diff
@@ -1,14 +1,19 @@
...
+  WithContextMenuProps,
+} from '@patternfly/react-topology';
 import { BaseEdge } from '@console/dev-console/src/components/topology';
 
 import './TrafficLink.scss';
 
 type TrafficLinkProps = {
   element: Edge;
-};
+} & WithContextMenuProps;
```
Comment (christianvogt) [inline frontend/packages/knative-plugin/src/topology/components/edges/TrafficLink.tsx:14]: Changes here aren't necessary if you type BaseEdge correctly.

Comment (karthikjeeyar) [reply frontend/packages/knative-plugin/src/topology/components/edges/TrafficLink.tsx]: updated

File: frontend/packages/knative-plugin/src/actions/add-subscription.tsx (hunk 3 above)
Comment (karthikjeeyar) [reply frontend/packages/knative-plugin/src/actions/add-subscription.tsx]: Yes, i have removed it.

**Comment by christianvogt** [comment]
@karthikjeeyar ran into an issue where if I set only the value of a filter and then submit the form, I get an error and then the filters key, value, and Add button go missing. 

image

**Comment by christianvogt** [comment]
The rest seems to function well.

File: frontend/packages/knative-plugin/src/actions/getEventPubSubActions.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+import { K8sKind } from '@console/internal/module/k8s';
+import { getKebabActionsForKind } from '../utils/kebab-actions';
+import { EventingBrokerModel } from '../models';
+
+export const getEventPubSubActions = (model: K8sKind) => {
+  let commonActions = Kebab.factory.common;
+
+  if (model.kind === EventingBrokerModel.kind) {
+    commonActions = Kebab.factory.common.filter((action) => {
+      return !['Delete', 'Edit'].includes(action.name);
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/actions/getEventPubSubActions.ts:12]: @karthikjeeyar will this filter out `delete` and `edit` for broker?

File: frontend/packages/knative-plugin/src/components/pub-sub/PubSub.tsx [hunk 9]
```diff
@@ -0,0 +1,87 @@
+import * as React from 'react';
+import { Formik, FormikValues, FormikHelpers } from 'formik';
+import { K8sResourceKind, k8sCreate } from '@console/internal/module/k8s';
+import PubSubModal from './PubSubModal';
+import { EventingBrokerModel, EventingTriggerModel, EventingSubscriptionModel } from '../../models';
+import { pubsubValidationSchema } from './pubsub-validation-utils';
+
+export interface PubSubProps {
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/components/pub-sub/PubSub.tsx:8]: nit: can remove export here

File: frontend/packages/knative-plugin/src/topology/create-connector-utils.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+  EventingSubscriptionModel,
+  EventingTriggerModel,
+} from '../models';
+
+export const getKnativeContextMenuAction = (
+  graphData: GraphData,
+  menu: MenuOptions,
+  connectorSource?: Node,
+): MenuOptions => {
+  if (isEventingChannelResourceKind(connectorSource?.getData().data.kind)) {
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/topology/create-connector-utils.ts:23]: nit : call `isEventingChannelResourceKind` only if `connectorSource` is defined

File: frontend/packages/knative-plugin/src/actions/getEventPubSubActions.ts (hunk 8 above)
Comment (karthikjeeyar) [reply frontend/packages/knative-plugin/src/actions/getEventPubSubActions.ts:12]: Yes, this is to filter those two common actions

File: frontend/packages/knative-plugin/src/topology/create-connector-utils.ts (hunk 10 above)
Comment (karthikjeeyar) [reply frontend/packages/knative-plugin/src/topology/create-connector-utils.ts:26]: Fixed all the comments, including filter section issue while the error message is set.

File: frontend/packages/knative-plugin/src/actions/getEventPubSubActions.ts (hunk 8 above)
Comment (invincibleJai) [reply frontend/packages/knative-plugin/src/actions/getEventPubSubActions.ts:12]: we don't need to filter karthik , we decided to have these as common actions based on user permissions i.e edit label/annotations are essentially edit broker

File: frontend/packages/knative-plugin/src/components/pub-sub/PubSub.tsx [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  close,
+  target = { metadata: { name: '' } },
+}) => {
+  const {
+    apiVersion: sourceApiVersion,
+    kind: sourceKind,
+    metadata: { namespace, name: sourceName },
+  } = source;
+  const {
+    apiVersion: tApiVersion = '',
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/components/pub-sub/PubSub.tsx:27]: nit: let's call it `targetApiVersion`

File: frontend/packages/knative-plugin/src/components/pub-sub/PubSubModal.tsx [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  filterEnabled,
+  labelTitle,
+  handleSubmit,
+  cancel,
+  isSubmitting,
+  status,
+  dirty,
+  errors,
+}) => {
+  return (
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/components/pub-sub/PubSubModal.tsx:33]: nit: can do a implicit return

File: frontend/packages/knative-plugin/src/topology/create-connector-utils.ts [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+  menu: MenuOptions,
+  connectorSource?: Node,
+): MenuOptions => {
+  if (!connectorSource) {
+    return menu;
+  }
+  if (isEventingChannelResourceKind(connectorSource?.getData().data.kind)) {
+    return [addSubscription(EventingSubscriptionModel, connectorSource.getData().resource)];
+  }
+  switch (connectorSource?.getData().data.kind) {
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/topology/create-connector-utils.ts:29]: nit: we can avoid `connectorSource?` now

**Comment by rachael-phillips** [comment]
LGTM!

**Comment by invincibleJai** [comment]
Verified , works as expected!!

**Pull request merged**